### PR TITLE
fix(spectrum): reproject only the active waterfall stream while panning (#3700)

### LINKED ISSUE #3700: Regression (#3668): waterfall pan choppy on wide displays — unconditional kiwi reproject deep-copies the image each pan step
## What happened?

Dragging the waterfall to pan (horizontal frequency scroll) is choppy again on high-resolution / wide displays, with the UI thread stalling ~60–80 ms per pan step. By inspection this traces to **#3668 (clean-room KiwiSDR integration)**, merged 2026-06-20 — the same day. It re-introduces, by a different mechanism, the per-pan-step waterfall cost that **#3578** had removed (which itself fixed the same class of choppiness, #3575, on wide displays).

### Root cause

#3578 made per-pan reproject cheap: the waterfall **history** image is no longer globally reprojected — each history row records its own capture frame and only the visible viewport is remapped.

#3668 changed `SpectrumWidget::handleWaterfallFrequencyFrameChange()` from a single `reprojectWaterfall()` into an **unconditional native + kiwi double reproject**, each wrapped in a waterfall-stream-state `save`/`restore`:

```cpp
reprojectStream(false);   // native
reprojectStream(true);    // kiwi — runs every pan step even with no KiwiSDR receiver
```

On **every pan mouse-move, even when KiwiSDR is inactive**, this costs two things:

1. A second, wasted reproject pass for a stream with no on-screen data.
2. `saveCurrentWaterfallStreamState()` does `state.waterfall = m_waterfall` — a QImage assignment, so the buffer is now **shared (copy-on-write)**. The next `m_waterfall.fill(Qt::black)` inside `rebuildWaterfallViewportForFrame()` therefore **detaches = deep-copies the entire waterfall image** (large on wide/high-DPI displays) on every pan step. That deep-copy is exactly the per-step cost #3578 had eliminated.

### Measured (`aether.perf`, same telemetry as #3578)

i9-13900K / RTX 4090, FLEX-6600, wide high-resolution panadapter, D3D11/GPU path. **Same binary**; one env var (`AETHER_WF_KIWI_ALWAYS`) toggles **only** the second pass. During a continuous hand pan:

| metric | with #3668 double pass | active-stream only |
|---|---|---|
| `wfUpdateP95Ms` | **63–68 ms** | **0.1 ms** |
| `panAgeP95Ms` (display lag) | 150–174 ms | 22–32 ms |
| `inputP95Ms` | 19–24 ms | 0.2–3.6 ms |
| `uiLagMaxMs` | 75–85 ms | 23–41 ms |
| total `PerfStall` (comparable pan) | 931 | 15 |

`fftRestarts`/`wfRestarts` = 0 and packet loss = 0 in both runs — the radio/stream is not involved. `renderP95` ~7 ms in both — the GPU render is not the bottleneck. The cost is purely the second pass + the deep-copy.

## What did you expect?

Pan to stay smooth (as after #3578), regardless of whether a KiwiSDR receiver is connected.

## Steps to reproduce

1. Connect to a FLEX radio; open a panadapter with a waterfall on a wide / high-resolution display.
2. Grab the waterfall and pan (drag horizontally) continuously for a few seconds.
3. Panning is choppy/laggy. With `aether.perf` enabled (Help → Support → "Performance"), `wfUpdateP95Ms` spikes to ~60–80 ms and `panAgeP95Ms` to ~150–170 ms during the drag.

A/B from one binary: launch with `AETHER_WF_KIWI_ALWAYS=1` → stall reproduces; without it (active-stream-only) → stall gone.

## AetherSDR version

26.6.3

## Radio model & firmware

FLEX-6600

## Operating system

Windows

## OS version and hardware

Windows 11 (build 26200), Qt 6.8.3 msvc2022_64; i9-13900K / RTX 4090, wide high-resolution panadapter.

---

Fix PR incoming (reproject only the active stream; remap the other on stream-switch). Couldn't self-assign — this account has no triage permission here.

### PR DESCRIPTION
**Pull request by svabi79**
## Summary

Fixes #3700. Panning the waterfall regressed on wide / high-DPI displays after **#3668** (KiwiSDR integration): every pan mouse-move ran an unconditional **native + kiwi** double reproject, even with no KiwiSDR receiver connected. The second pass is wasted work, and its waterfall-stream-state save/restore leaves `m_waterfall` copy-on-write–shared, so the next `fill()` in `rebuildWaterfallViewportForFrame()` **deep-copies the whole waterfall image** every pan step — 60–80 ms of UI-thread stall per step. This re-introduced, by a different path, the per-pan-step cost that **#3578** (#3575) had removed.

This PR reprojects **only the active stream** per pan, and remaps the other stream to the current frequency frame when it next becomes visible (`setKiwiSdrWaterfallActive`) — exact, because each history row carries its own capture frame (#3578). `AETHER_WF_KIWI_ALWAYS=1` restores the old double pass for A/B.

## How we got here (approach)

Came at this from interest in the panadapter-perf topic — same lane as #3578/#3617. The path mattered, so here it is:

1. **Measured first, didn't guess.** `aether.perf` showed the pan stall is client-side: `wfUpdateP95` 0.1 ms idle → ~68 ms during pan, `panAgeP95` ~160 ms, 200+ main-thread stalls of 60–83 ms — while `renderP95` stayed ~7 ms and `fftRestarts`/`wfRestarts`/loss were all 0. So **not** the GPU render, and **not** the radio/retune (our first hypothesis — the data disproved it).
2. **Bisected by inspection.** The cost lands in `rebuildWaterfallViewportForFrame` (fill + remap every visible row) — but #3578 had made that acceptable. `git log` on the pan path since #3578 surfaced exactly one suspect: **#3668** replaced the single `reprojectWaterfall()` with `reprojectStream(false) + reprojectStream(true)` (merged the same day the choppiness appeared).
3. **Two mechanisms.** The unconditional second (kiwi) pass; and the QImage COW detach the stream-state save/restore triggers on the next `fill()` — a full deep-copy of the (large) waterfall image, per pan step.
4. **Proved it with a one-variable A/B** — same binary, `AETHER_WF_KIWI_ALWAYS` toggling *only* the second pass:

| metric (hand pan) | #3668 double pass | this PR (active only) |
|---|---|---|
| `wfUpdateP95Ms` | 63–68 ms | **0.1 ms** |
| `panAgeP95Ms` | 150–174 ms | **22–32 ms** |
| `inputP95Ms` | 19–24 ms | **0.2–3.6 ms** |
| `uiLagMaxMs` | 75–85 ms | **23–41 ms** |
| total `PerfStall` | 931 | **15** |

The single switch flips it — the second pass *is* the regression, reprojecting only the active stream *is* the fix.

## The change

- **`handleWaterfallFrequencyFrameChange()`** — reproject only the active stream (`reprojectStream(m_kiwiSdrWaterfallActive)`): no stream swap, so no COW-share and no per-step deep-copy.
- **`setKiwiSdrWaterfallActive()`** — after restoring the switched-in stream's state, `rebuildWaterfallViewportForFrame(m_centerMhz, m_bandwidthMhz)` aligns it to the current frame the moment it's shown. This is the lazy counterpart to the per-pan reproject the inactive stream used to receive, so toggling streams stays seamless.
- **`AETHER_WF_KIWI_ALWAYS=1`** keeps the old unconditional double pass available for verification.

## Constitution principle honored

**Principle XI — Fixes Are Demonstrated.** Root cause and fix both confirmed by `aether.perf` before/after (table above), same binary, one variable changed.

## Test plan

- [x] Local build passes (Windows, Qt 6.8.3 msvc2022_64, GPU path on)
- [x] Before/after measured with the gated `aether.perf` telemetry (two runs each)
- [ ] Existing tests pass (CI)
- [x] Reproduction steps documented in #3700
- [ ] ⚠️ **KiwiSDR-active path not hardware-tested** — I have no KiwiSDR receiver, so the measured path is FLEX (active = native). The stream-switch remap reuses the existing `rebuildWaterfallViewportForFrame`; a KiwiSDR-equipped user toggling streams mid-pan would confirm the switch stays aligned. Flagging for maintainer / KiwiSDR review.

## Checklist

- [x] Commits are signed (SSH)
- [x] No new flat-key `AppSettings` calls (N/A)
- [x] Code is clean-room (Principle IV) — not decompiled / reverse-engineered
- [x] All meter UI uses `MeterSmoother` (N/A)
- [x] Documentation updated if user-visible behavior changed (N/A — pan is visually unchanged, just smooth again)
- [x] Security-sensitive changes reference a GHSA if applicable (N/A)

### CODE REVIEW

**Review by ten9876 — APPROVED** [review]
Approving. Verified the diagnosis and fix against the code:

- **Regression origin confirmed** — `git log -S "reprojectStream(true)"` pins the unconditional native+kiwi double pass to #3668. It runs on every pan step with no KiwiSDR guard, so it regressed the native panadapter for *all* users (the author measured it FLEX-only, active=native).
- **COW deep-copy mechanism is real** — the inactive-stream reproject round-trips through `save/restoreCurrentWaterfallStreamState`, leaving `m_waterfall` implicitly shared so the next `fill()` detaches the full image per pan step. Reprojecting only the active stream hits the `visibleStream` early-return in `begin/endWaterfallStreamWrite` — no share, no detach.
- **Lazy remap is exact** — each history row carries its own capture frame (#3578) and `rebuildWaterfallViewportForFrame` remaps per-row, so deferring the inactive stream's reprojection to `setKiwiSdrWaterfallActive` is equivalent. No stale-smoothing regression (kiwi smoothing is in fixed aperture bin-space, invariant to a display pan; ingest clears on frame change).
- Minimal (+32/-6, one file), excellent Principle XI demonstration (measure → bisect → one-variable A/B), env-gated escape hatch retained, no UX change.

Known caveat (author-flagged, accepted): the kiwi-active / stream-switch-mid-pan branch is reasoned + traced but not hardware-tested. Mechanism is symmetric to the verified native path and env-reversible.

Thanks @svabi79 — exemplary write-up.
